keps/0029-operator-dependencies.md

```diff
@@ -0,0 +1,281 @@
+
+## Proposal
+
+KUDO operators **already have** a mechanism to deal with installation dependencies called [plans, phases, and steps](https://kudo.dev/docs/developing-operators/plans.html#overview) with serial or parallel execution strategy. This mechanism is already powerful enough to express any dependency hierarchy including transitive dependencies (see more about this in [implementation details](#implementation-details)). The core of this proposal is to reuse this mechanism and extend it with the ability to install operators.
+
+### Implementation details
+
+#### Operator Task
+
+A new task kind `Operator` is introduced which extends `operator.yaml` with the ability to install dependencies. Let's take a look at the Kafka+Zookeeper example:
+
+```yaml
+apiVersion: kudo.dev/v1beta1
+name: "kafka"
+operatorVersion: "1.3.1"
+kudoVersion: 0.12.0
+kubernetesVersion: 1.14.8
+appVersion: 2.5.0
+url: https://kafka.apache.org/
+tasks:
+  - name: zookeeper-operator
+    kind: Operator
+    spec:
+        package: zookeeper
+
+  ...
+
+plans:
+  deploy:
+    strategy: serial
+    phases:
+      - name: deploy-zookeeper
+        strategy: serial
+        steps:
+          - name: zookeeper
+            tasks:
+              - zookeeper-operator
+      - name: deploy-kafka
+        strategy: serial
+        steps:
+          ...
+```
+
+The `zookeeper-operator` task specification is equivalent to `kudo install zookeeper` CLI command which installs the Zookeeper package from the official repo. Here is a complete `Operator` task specification:
+
+```yaml
+tasks:
+- name: demo
+  kind: Operator
+  spec:
+    package: # required, either repo package name, local package folder or an URL to package tarball
+    repo: # optional, name of local repository configuration to use
+    appVersion: # optional, a specific app version in the official repo, defaults to the most recent one
+    operatorVersion: # optional, a specific operator version in the official repo, defaults to the most recent one
+    instanceName: # optional, the instance name
+```
+
+As you can see, this closely mimics the `kudo install` CLI command [options](https://github.com/kudobuilder/kudo/blob/master/pkg/kudoctl/cmd/install.go#L56) because at the end the latter will be executed to install the operator. We omit `parameters` and `parameterFile` options at this point as they are discussed in detail [below](#dependencies-parametrization).
+
+#### Deployment
+
+##### Client-Side
+
+Upon execution of `kudo install ...` command with the above operator definition, CLI will:
+
+1. Collect all operator dependencies by analyzing the `deploy` plan of the top-level operator and compose a list of all operator dependencies (tasks with the kind `Operator`) including **transitive dependencies**
+2. Install all collected operators **skipping the instances** (same as `kudo install ... --skip-instance`). Note that since we do not create instances here we can install them in any order
```

Comment:
```suggestion
2. Install all collected operator versions **skipping the instances** (same as `kudo install ... --skip-instance`). Note that since we do not create instances here we can install them in any order
```